### PR TITLE
fix(3d): reject a sidecar rotate that inherits a frame-mismatched packaged offset

### LINKED ISSUE #4636: 3d: LCSC packaged offset is calibrated for its sibling rotate — per-field merge lets a sidecar rotate inherit a mismatched offset
## Context

PR #4631 (issue #4584, merged as `6623f59f`) added per-part 3D-body transforms for the LCSC/EasyEDA model tier, with two override sources resolved in `models3d._resolve_lcsc`. Per `src/kicad_tools/pcb/lcsc_model_transforms.py:39-46`:

> **Precedence** (resolved in `models3d._resolve_lcsc`, first non-`None` wins, **merged per field**):
> 1. A per-board `lcsc_models.json` sidecar entry in object form
> 2. This packaged table
> 3. Identity

The same module documents the frame semantics (`:30-37`):

> Because `offset` is applied *after* `rotate`, the `offset` here is a post-rotation translation — so when calibrating a part, **set `rotate` first, render, and only then measure `offset`**.

## The latent defect

Those two rules interact badly. Because the merge is **per field**, a board sidecar can override `rotate` alone and silently inherit the packaged table's `offset` — an offset that was measured *under a different rotation*.

Concretely: a packaged entry calibrated as `rotate=(0,0,-90), offset=(1.5,-0.25,0)` has an X/Y offset that is only meaningful in the −90° frame. A sidecar saying `{"lcsc": "C…", "rotate": [0,0,90]}` inherits that `(1.5,-0.25,0)` and applies it in the +90° frame, where it points somewhere else entirely. The result is a body that is *almost* right — plausibly placed, quietly wrong — which is the worst failure mode for a transform whose only validation is a human looking at a render.

The Z component is unaffected (it is rotation-invariant for a Z-axis rotation); this is an X/Y problem specifically.

## Exposure today: zero

No packaged-table entry carries an `offset` yet — the single entry, `C444929`, is `rotate`-only, and its provenance notes explicitly record *"A front-side render confirms the body seats on the board surface, so no offset is needed."* So there is currently no pair for a sidecar to split. This is a trap laid for the second calibration, not an active bug.

## Options

1. **Couple the fields on override.** If a sidecar specifies `rotate`, do not inherit a packaged `offset` (require the sidecar to restate it, or reset it to zero). Trades a small surprise for a larger one.
2. **Reject the split explicitly.** Error out when a sidecar overrides `rotate` while the packaged table supplies a non-zero X/Y `offset`, naming both sources. Loud beats subtle, and matches the module's existing posture of rejecting unknown keys rather than ignoring them.
3. **Document only.** A line beside the "merged per field" framing warning that `offset` is calibrated for its sibling `rotate`. Cheapest; relies on the next calibrator reading it.

Given the module's stated philosophy — *"only a render can tell you whether a number is right"* and provenance is structurally enforced rather than merely requested — option 2 seems most in keeping: an inherited-offset-under-a-new-rotation is exactly an unverified number.

## Acceptance criteria

- [ ] A sidecar overriding `rotate` cannot silently inherit a packaged X/Y `offset` calibrated under a different rotation — whether by coupling, by rejection, or by documented contract
- [ ] Whatever rule is chosen is stated at the precedence docstring in `lcsc_model_transforms.py`, not only in a commit message
- [ ] A test covers the split case (sidecar `rotate` + packaged `offset`)
- [ ] Behavior for the two existing sidecars (board 03 → `C50950`, board 06 → `C444929`) is unchanged

## Related

- #4584 / PR #4631 — introduced the packaged table and the per-field merge
- #4585 — board 03/06/07 3D regen, the first real consumer of these transforms
- #4586 — `add-3d-models` is insert-only, so existing `(model …)` nodes must be stripped by hand before regen

*Surfaced by the Judge review of PR #4631 during post-crash sweep recovery, 2026-08-04. Citations verified against `main` @ `6623f59f`.*

---

# Curator Enhancement (2026-08-04)

> All file/line citations and counts in this section were verified against `origin/main` @ `473de1f8` (which contains `6623f59f`) on **2026-08-04**. Re-verify before relying on any line number — `models3d.py` is also the target of #4586.

## Verification of the report

Confirmed, and the report's framing is accurate:

- The merge site is `src/kicad_tools/pcb/models3d.py:746-753`, inside `make_library_resolver._resolve_lcsc`. The two merge lines are literally independent:
  ```python
  rotate = sidecar_rotate if sidecar_rotate is not None else packaged_rotate
  offset = sidecar_offset if sidecar_offset is not None else packaged_offset
  ```
  Nothing couples them, and nothing validates the pair.
- **Exposure is still zero as of `473de1f8`, 2026-08-04**: `LCSC_MODEL_TRANSFORMS` has exactly one entry (`C444929`, `rotate`-only, no `offset`), and both committed sidecars are in **bare-string form** — `boards/03-usb-joystick/lcsc_models.json` is `{"Module:Joystick_Analog": "C50950"}` and `boards/06-diffpair-test/lcsc_models.json` is `{"Connector_PCIE:PCIE_Mini_Edge": "C444929"}`. Neither carries a `rotate`, so neither can trigger the split. Acceptance criterion 4 ("behavior for the two existing sidecars unchanged") is satisfied *by construction* under any of the three options — but should still be asserted by a test so it stays true.
- `C50950` (board 03) has **no** packaged entry at all, so it resolves to identity.

## Suspected cause (verified by reading, not by render)

Not a code defect — a **contract gap**. The per-field merge and the post-rotation `offset` frame were each designed correctly in isolation (`lcsc_model_transforms.py:30-46`); their interaction was never specified. There is no observed misrender to reproduce, because no packaged `offset` exists yet to be inherited. Treat this as hardening a trap, not fixing a live bug — a Builder who instruments this and reports "the split cannot occur today" has confirmed the premise, not refuted it.

## Recommended approach: Option 2 (reject the split), narrowly scoped

Option 2 is the right call for the reasons the reporter gives, but it must be **scoped precisely** or it will either over-reject (breaking the existing per-field tests) or under-reject (missing the general-rotation case). Recommended predicate — raise only when **all** of these hold:

1. `sidecar_rotate is not None` — the sidecar overrides `rotate`;
2. `sidecar_offset is None` — the `offset` is *inherited*, not restated;
3. a packaged entry exists and its `offset is not None`;
4. `sidecar_rotate != pkg_rot`, where `pkg_rot = packaged.rotate if packaged.rotate is not None else (0.0, 0.0, 0.0)` — a sidecar that redundantly *restates* the packaged rotation has not changed the frame, so its inherited offset is still valid;
5. at least one **invalidated** component of the packaged `offset` is non-zero, where the invalidated set is:
   - `{X, Y}` when `sidecar_rotate` and `pkg_rot` differ **only** in their Z component (a Z-axis rotation leaves Z invariant — the reporter's observation), or
   - `{X, Y, Z}` when they differ in X and/or Y (an off-Z rotation tilts the body, so Z is no longer invariant either).

Exact float comparison is acceptable for condition 4 — both sides are authored literals, never the result of arithmetic.

**Why this scoping matters:** the two existing merge tests both use a **Z-only** packaged offset `(0.0, 0.0, 1.25)` with rotations that differ only in Z (`tests/test_lcsc_models.py:875-906`). Under the predicate above they stay green with no edit. A naive "any sidecar rotate + any packaged offset ⇒ error" rule would break both and would be over-strict for the legitimate "STEP origin is off the board plane" knob, which is exactly what the Z component is documented to be (`lcsc_model_transforms.py:110-118`).

**The symmetric case is NOT a defect and must keep working**: a sidecar overriding only `offset` while inheriting a packaged `rotate` is fine — the author necessarily rendered *with* the packaged rotation active (they did not override it), so their measured offset is already in the right frame. `test_merge_is_per_field_in_the_other_direction` must stay green.

### Escape hatch (no code change needed to unblock a legitimate override)

An explicit `"offset": [0, 0, 0]` in the sidecar sets `sidecar_offset` to a non-`None` `(0.0, 0.0, 0.0)`, which wins the merge and suppresses the packaged value (see the merge line above; `test_explicit_zero_triples_are_indistinguishable_from_identity` confirms the emitted output is unchanged). So the rejection is always escapable by *acknowledging* the offset — which is precisely the intent: force the calibrator to make a decision instead of inheriting one silently. Say so in the error message.

### Where the rule should live

Put the rule in `src/kicad_tools/pcb/lcsc_model_transforms.py`, not inline in `models3d.py` — that module already owns the frame semantics and the precedence docstring (acceptance criterion 2 requires the rule be stated *there*), and it is import-light. Suggested shape:

```python
def resolve_merged_transform(
    c_number: str,
    lib_id: str,
    sidecar_rotate: Triple | None,
    sidecar_offset: Triple | None,
) -> tuple[Triple | None, Triple | None]:
    """Merge the sidecar over the packaged table, rejecting a frame-mismatched
    inherited offset.  Raises ValueError on the split-calibration case."""
```

`_resolve_lcsc` then replaces its two merge lines with one call. This keeps the rule unit-testable without constructing a resolver, and keeps `models3d.py`'s diff to ~3 lines — which matters because #4586 also edits that file (see Coordination below).

**The CLI error path already exists**: `src/kicad_tools/cli/pcb_add_3d_models.py:94-110` wraps `add_model_refs` in `except Exception` and returns exit 1, printing `Error: {e}` to stderr for `--format text` and `{"error": ...}` for `--format json`. A `ValueError` raised from `_resolve_lcsc` surfaces cleanly through that path with no CLI change — matching how a malformed sidecar is already handled (`:69-76`). No new plumbing required.

**Error-message content** — name both sources, because "naming both sources" is the whole point of choosing loud over subtle. Available at the raise site: the `lib_id`, the C-number, the packaged `rotate`/`offset` triples, the sidecar `rotate`, and the packaged entry's `provenance` (board / refdes / `verified` date), which is the direct evidence of *which frame the offset was measured in*. Include the remedy: restate `offset` in the sidecar (`[0, 0, 0]` if a re-render shows none is needed), or drop the sidecar `rotate`.

**Not available at the raise site**: the sidecar file path. `LcscModelEntry` (`src/kicad_tools/pcb/lcsc_models.py:142-155`) carries only `lcsc`/`rotate`/`offset`; the path is consumed by `load_lcsc_mapping` and discarded. Name the `lib_id` instead — do **not** plumb the path through as part of this issue (scope creep into `LcscMapping`'s public type).

## Docs/comment sites that assert the superseded rule

The current "merged per field, full stop" contract is stated in **six** places (enumerated by `git grep -n -i "per field\|per-field" -- src docs` @ `473de1f8`, 2026-08-04; a seventh hit in `tests/test_net_class_serialization.py:248` is unrelated). Every one of them must be updated, or the codebase will contradict itself:

| File | Line | What it says |
|---|---|---|
| `src/kicad_tools/pcb/lcsc_model_transforms.py` | 39-46 | The **canonical** precedence docstring — AC 2 targets this one specifically |
| `src/kicad_tools/pcb/lcsc_models.py` | 30 | Module docstring precedence summary |
| `src/kicad_tools/pcb/lcsc_models.py` | 147-150 | `LcscModelEntry` docstring: "a sidecar that overrides only `rotate` does not suppress a packaged `offset`" |
| `src/kicad_tools/pcb/models3d.py` | 646 | `make_library_resolver` docstring |
| `src/kicad_tools/pcb/models3d.py` | 746-748 | The inline comment directly above the merge |
| `docs/guides/lcsc-3d-models.md` | 51, 59-60 | Precedence table intro + "Merging is **per field**: a sidecar entry that overrides only `rotate` still inherits the packaged `offset`, and vice versa." |

The guide should also gain a sentence in **Frame semantics** (`docs/guides/lcsc-3d-models.md:96-112`), next to the existing "set `rotate` first, render, and only then measure `offset`" — that is where a calibrator will actually be reading.

## Explicitly out of scope

- Changing `--refresh`/rewrite behavior (that is #4586).
- Regenerating any board artifact or render (that is #4585). **No file under `boards/*/output/` may change in this PR.**
- Adding any new packaged-table entry, or altering `C444929`.
- Plumbing the sidecar file path into `LcscModelEntry` / `LcscMapping`.
- The symmetric `offset`-override-inheriting-`rotate` case (not a defect — see above).

## Refined acceptance criteria

Supersedes the original list above; items 1-4 are the originals, sharpened.

- [ ] A sidecar overriding `rotate` cannot silently inherit a packaged `offset` whose invalidated components are non-zero — implemented as a loud `ValueError` (Option 2), scoped by the five-condition predicate above
- [ ] The rule is stated at the precedence docstring in `src/kicad_tools/pcb/lcsc_model_transforms.py`, and every one of the six sites in the table above is updated so no doc/comment still asserts the unqualified per-field contract
- [ ] Tests cover the split case (sidecar `rotate` + packaged X/Y `offset`) **and** the three cases that must remain legal: Z-only offset under a Z-only rotation delta, redundant identical `rotate`, and an explicitly restated sidecar `offset`
- [ ] Behavior for the two committed sidecars (board 03 -> `C50950`, board 06 -> `C444929`) is unchanged; both are bare-string form and neither raises
- [ ] The error message names the `lib_id`, the C-number, the packaged `rotate`/`offset`, the sidecar `rotate`, the packaged entry's provenance (board/refdes/date), and the remedy
- [ ] `test_merge_is_per_field_not_all_or_nothing` and `test_merge_is_per_field_in_the_other_direction` (`tests/test_lcsc_models.py:875-906`) still pass **without modification** — if either needs editing, the predicate is over-strict
- [ ] No file under `boards/*/output/` is modified

## Affected Files

- `src/kicad_tools/pcb/lcsc_model_transforms.py` — **primary**: add the merge/validation helper (`resolve_merged_transform` or equivalent) that rejects the frame-mismatched inherited offset; update the precedence docstring (`:39-46`) to state the new rule
- `src/kicad_tools/pcb/models3d.py` — replace the two independent merge lines in `_resolve_lcsc` (`:746-753`) with a call to the new helper; update the inline comment (`:746-748`) and the `make_library_resolver` docstring (`:646`). **Keep this diff minimal — collides with #4586** (see Coordination)
- `src/kicad_tools/pcb/lcsc_models.py` — docstring-only: correct the module precedence summary (`:30`) and the `LcscModelEntry` docstring (`:147-150`), both of which currently assert the unqualified per-field contract
- `tests/test_lcsc_models.py` — add the split-case + legal-case tests (suggested: a new `TestSplitCalibrationGuard` class beside the existing `TestTransformPrecedence` at `:834`); do **not** edit the two existing per-field tests
- `tests/test_pcb_add_3d_models.py` — one CLI-surface test: the `ValueError` exits 1 and prints the message under `--format text` and `--format json`
- `docs/guides/lcsc-3d-models.md` — correct the per-field statement (`:51`, `:59-60`) and add the calibrator-facing warning in **Frame semantics** (`:96-112`)
- `boards/03-usb-joystick/lcsc_models.json` — **read-only reference** (bare-string form; must remain unchanged, AC 4)
- `boards/06-diffpair-test/lcsc_models.json` — **read-only reference** (bare-string form; must remain unchanged, AC 4)

All eight paths verified present on `origin/main` @ `473de1f8` (curator pre-flight, 2026-08-04).

## Test Plan

- [ ] **Automated — the defect**: sidecar `{"lcsc": "C…", "rotate": [0,0,90]}` over a packaged entry with `rotate=(0,0,-90), offset=(1.5,-0.25,0)` raises `ValueError`; assert the message contains the lib_id, the C-number, both triples, and the provenance date
- [ ] **Automated — off-Z rotation invalidates Z too**: sidecar `rotate=(0,15,0)` over packaged `rotate=(0,0,0), offset=(0,0,1.25)` raises (the Z-only offset is no longer rotation-invariant)
- [ ] **Automated — legal 1 (Z-invariant)**: sidecar `rotate=(0,0,180)` over packaged `rotate=(0,0,-90), offset=(0,0,1.25)` resolves silently, `offset` inherited — this is the existing `test_merge_is_per_field_not_all_or_nothing`, must pass unmodified
- [ ] **Automated — legal 2 (redundant restatement)**: sidecar `rotate` exactly equal to the packaged `rotate`, packaged `offset=(1.5,-0.25,0)` — resolves silently
- [ ] **Automated — legal 3 (acknowledged)**: sidecar supplies both `rotate` and `offset` (including an explicit `[0,0,0]`) — resolves silently, sidecar `offset` wins
- [ ] **Automated — legal 4 (symmetric case)**: sidecar `offset`-only inheriting a packaged `rotate` — existing `test_merge_is_per_field_in_the_other_direction`, must pass unmodified
- [ ] **Automated — legal 5 (bare string)**: bare-string sidecar over a packaged `rotate`+`offset` pair — resolves silently (the normal calibrated-pair path)
- [ ] **Automated — CLI surface**: `kct pcb add-3d-models` exits 1 and prints the message on both `--format text` (stderr) and `--format json` (`{"error": ...}`)
- [ ] **Automated — committed sidecars**: `load_lcsc_mapping` on `boards/03-usb-joystick/lcsc_models.json` and `boards/06-diffpair-test/lcsc_models.json` resolves without raising (guards AC 4 against future table additions)
- [ ] **Manual verification**: `uv run kct pcb add-3d-models boards/06-diffpair-test/output/diffpair_test_routed.kicad_pcb --lcsc-models boards/06-diffpair-test/lcsc_models.json --dry-run` — output identical to `main`, no error. **Run this in a scratch worktree or with `--dry-run` only; it must not dirty `boards/*/output/`**
- [ ] **Edge cases**: packaged entry with `offset` but no `rotate` (`pkg_rot` defaults to identity, still rejects a sidecar rotate against a non-zero X/Y offset); packaged entry with `provenance=None` (must not `AttributeError` while building the message); a `-0.0` offset component (treat as zero)
- [ ] `uv run pytest tests/test_lcsc_models.py tests/test_pcb_add_3d_models.py`, plus `uv run ruff check .` / `uv run ruff format --check .` / `uv run mypy src/`

## Coordination with the sibling issues in this sweep

- **#4586** (`add-3d-models --refresh`) — **direct file collision on `src/kicad_tools/pcb/models3d.py`**, and likely on `tests/test_pcb_add_3d_models.py` and `docs/guides/lcsc-3d-models.md`. #4586 rewrites the insert-only patch path (`add_model_refs_to_text`, `:835+`) and the CLI flag surface; this issue touches `_resolve_lcsc` (`:706-771`) plus three doc-comment lines. The regions are disjoint but the files are shared — **schedule these in different waves**, or land this one first (it is the smaller diff and #4586's `--refresh` would otherwise need rebasing over the merge-helper extraction).
- **#4585** (regen boards 03/06/07 + gallery) — **no source-file overlap**; it edits `boards/*/output/` artifacts and renders. It is, however, the first real *consumer* of this merge path. Functionally unaffected today (both sidecars are bare-string, and `C444929` has no packaged `offset`), so it can run in parallel. If #4585 lands first and someone later adds a packaged `offset` for `C444929`, this guard is what keeps a subsequent board-local `rotate` override from inheriting it.
- **#4584 / PR #4631** (`6623f59f`) — merged; introduced the table and the per-field merge this issue constrains.

## Duplicate / related-open-work check

`./.loom/scripts/check-duplicate.sh --include-merged-prs --issue 4636` (2026-08-04) returned exit 1 with a single `DUPLICATE_FOUND` hit: **PR #4631 at 21% similarity** — that is the merged PR whose Judge review *surfaced* this issue, not a duplicate. No `RELATED_OPEN_WORK` block was returned, so no open issue argues for a different spec here. Not a duplicate; proceeding.

### PR DESCRIPTION
**Pull request by rjwalters**
## Summary

A packaged LCSC transform's `offset` is a **post-rotation** translation, so it is only meaningful in the frame of the `rotate` it was calibrated under. The per-field sidecar/table merge in `models3d._resolve_lcsc` let a board sidecar override `rotate` alone and silently inherit that offset, applying an X/Y translation measured in a frame that no longer exists — a plausibly-placed, quietly-wrong body, which is the worst failure mode for a transform whose only real validation is a human looking at a render.

Implements **Option 2** (reject the split explicitly), scoped by the five-condition predicate from the curator enhancement so it does not over-reject the legitimate "STEP origin is off the board plane" Z knob.

Exposure today is still zero — `LCSC_MODEL_TRANSFORMS` has one `rotate`-only entry (`C444929`) and both committed sidecars are bare-string form — so this hardens a trap laid for the *second* calibration rather than fixing a live misrender.

## Changes

- **`src/kicad_tools/pcb/lcsc_model_transforms.py`** — new `resolve_merged_transform(c_number, lib_id, sidecar_rotate, sidecar_offset)` performs the merge and raises `ValueError` when **all** of: the sidecar overrides `rotate`; the offset is *inherited* (sidecar `offset is None`); a packaged entry supplies an `offset`; the sidecar rotation differs from the packaged one (a packaged entry with no `rotate` is treated as calibrated at identity); and at least one **invalidated** offset component is non-zero. Invalidated = `{X, Y}` for a Z-only rotation delta, `{X, Y, Z}` when the delta is off-Z (a tilt moves the body vertically, so Z stops being rotation-invariant). The precedence docstring now states the rule and the three merges that stay legal.
- **`src/kicad_tools/pcb/models3d.py`** — the two independent merge lines in `_resolve_lcsc` become one call to the helper; the inline comment and the `make_library_resolver` docstring are corrected. Deliberately kept to ~3 code lines inside `_resolve_lcsc` so #4586's later `--refresh` work on this file rebases cleanly.
- **`src/kicad_tools/pcb/lcsc_models.py`** — docstring-only: module precedence summary + `LcscModelEntry`, which both asserted the unqualified per-field contract (the `LcscModelEntry` example was flipped to the symmetric direction, which remains unconditionally true).
- **`docs/guides/lcsc-3d-models.md`** — new "Guard: a packaged `offset` belongs to its sibling `rotate`" subsection under the precedence table, plus a corollary sentence in **Frame semantics**, right beside "set `rotate` first, render, and only then measure `offset`" — where a calibrator will actually be reading.
- **Tests** — new `TestSplitCalibrationGuard` (13 cases) in `tests/test_lcsc_models.py` and `TestSplitCalibrationCliSurface` (3 cases) in `tests/test_pcb_add_3d_models.py`.

The error message names the `lib_id`, the C-number, the packaged `rotate`/`offset`, the sidecar `rotate`, the packaged entry's provenance (board / refdes / verified date), which invalidated components are stale, and both remedies. Escape hatch, as specified: restating `offset` in the sidecar — including an explicit `[0, 0, 0]` — wins the merge and resolves the error, so the guard forces a *decision* rather than blocking a legitimate override.

## Acceptance Criteria Verification

- [x] **Sidecar `rotate` cannot silently inherit a frame-mismatched packaged `offset`** — `ValueError` from `resolve_merged_transform`, five-condition predicate as specified. `test_sidecar_rotate_inheriting_an_xy_offset_raises`, `test_off_z_rotation_invalidates_a_z_only_offset_too`, `test_packaged_offset_without_a_rotate_is_calibrated_in_the_identity_frame`.
- [x] **Rule stated at the precedence docstring, and all six sites updated** — `git grep -n -i "per field\|per-field" -- src docs` now returns 10 hits across the 4 files (plus 2 unrelated C++ router hits); every one is qualified by the guard. Sites: `lcsc_model_transforms.py:40`, `lcsc_models.py:30,33,151,154`, `models3d.py:646,749`, `docs/guides/lcsc-3d-models.md:51,59,66`.
- [x] **Tests cover the split case and the cases that must remain legal** — Z-only offset under a Z-only rotation delta, redundant identical `rotate`, restated sidecar `offset` (both a real value and an explicit `[0,0,0]`), bare-string sidecar over a calibrated pair, symmetric `offset`-only override, no packaged entry, signed-zero offset components, and `provenance=None` (must not `AttributeError` while building the message).
- [x] **Behavior for the two committed sidecars is unchanged** — `test_committed_sidecars_never_trip_the_guard` loads `boards/03-usb-joystick/lcsc_models.json` and `boards/06-diffpair-test/lcsc_models.json` through `load_lcsc_mapping` and runs every entry through the helper. Guards AC 4 against a future table addition rather than relying on today's bare-string form.
- [x] **Error message names all required facts** — asserted componentwise in `test_sidecar_rotate_inheriting_an_xy_offset_raises` (lib id, C-number, both triples, board, refdes, date, `[0, 0, 0]` remedy) and `test_message_names_only_the_invalidated_nonzero_components`.
- [x] **`test_merge_is_per_field_not_all_or_nothing` and `test_merge_is_per_field_in_the_other_direction` pass without modification** — `git diff` touches no line in `tests/test_lcsc_models.py:875-906`; the only edit above the new class is the `resolve_merged_transform` import. Both are green.
- [x] **No file under `boards/*/output/` modified** — `git show --stat` lists only the 6 files above; `git status` in the worktree is clean after commit.

## Test Plan

- `uv run pytest tests/test_lcsc_models.py tests/test_pcb_add_3d_models.py tests/test_render_cmd.py` → **244 passed** (131 + 71 + 42; 16 of them new).
- `uv run ruff check .` → All checks passed. `uv run ruff format --check .` → 1731 files already formatted.
- `uv run python scripts/ci/check_mypy_baseline.py` → `OK: mypy reports 1473 error(s), all within the baseline (1474 allowed). No new type errors.` (The baseline file is deliberately left untouched — see the known `--update` native-env trap.)
- **Manual, per the issue's test plan**: `uv run kct pcb add-3d-models boards/06-diffpair-test/output/diffpair_test_routed.kicad_pcb --lcsc-models boards/06-diffpair-test/lcsc_models.json --dry-run --format json` → exit 0, no error, `git status -- boards/` clean.
- **CLI surface**: exit 1 with the message on stderr under `--format text` and as `{"error": ...}` under `--format json`; the target `.kicad_pcb` is byte-identical after a rejected run; the `[0, 0, 0]` escape hatch returns the run to exit 0 with the footprint patched.

## Notes

- `CHANGELOG.md` intentionally untouched (three-way contended in this wave).
- Rotation equality is compared **exactly**, per the curator's guidance — both sides are authored literals, never arithmetic results. The docstring records why that is the *right* test rather than merely an acceptable one: a sidecar writing `360` where the table writes `0` has stated a rotation it never rendered, which is the class of unverified number this guard exists to catch.
- `-0.0` is correctly benign (`-0.0 != 0.0` is `False`), covered by `test_signed_zero_offset_components_count_as_zero`.

Closes #4636

### CODE REVIEW

**Comment by rjwalters** [comment]
## Judge review: APPROVED

Option 2 implemented exactly to the curator's five-condition predicate, with the rule stated where AC 2 requires it (`lcsc_model_transforms.py` precedence docstring) and all six superseded per-field sites updated.

### Over-strictness tripwire (the acceptance criterion that mattered)

**PASS, unmodified.** `git diff origin/main...HEAD -U0 -- tests/test_lcsc_models.py` produces exactly two hunks — `@@ -31,0 +32 @@` (the `resolve_merged_transform` import) and `@@ -945,0 +947,203 @@` (the appended `TestSplitCalibrationGuard`). **No line in `tests/test_lcsc_models.py:875-906` is touched.** Both tripwires run green as-is:

```
uv run pytest tests/test_lcsc_models.py -k per_field  ->  2 passed, 129 deselected
```

They survive for the right reasons, not by luck: `test_merge_is_per_field_not_all_or_nothing` has a Z-only packaged offset `(0,0,1.25)` under a Z-only rotation delta, so `_invalidated_offset_axes` returns `("x","y")` and both are zero -> `stale` is empty -> no raise. `test_merge_is_per_field_in_the_other_direction` overrides only `offset`, so condition 1 (`sidecar_rotate is not None`) fails outright. The documented "STEP origin is off the board plane" Z knob is preserved.

### Other scrutiny points

- **Packaged entry with no `rotate`** — condition 4 is implemented as the spec words it: `effective_rotate = packaged_rotate if packaged_rotate is not None else _IDENTITY_ROTATE`, then compared. It does not fire *because* the rotate is absent; it fires only if the sidecar rotation differs from identity, which is a genuine frame change for a post-rotation offset. Correct.
- **All-zero invalidated components** — `stale` is built with `packaged_offset[index] != 0.0`; empty list short-circuits the raise. `-0.0` is benign for free (`-0.0 != 0.0` is `False`), and pinned by `test_signed_zero_offset_components_count_as_zero` rather than left to a reader's inference.
- **Off-Z deltas invalidate Z too** — `_invalidated_offset_axes` compares only rotation components 0 and 1; any X/Y difference widens the set to `("x","y","z")`. Covered by `test_off_z_rotation_invalidates_a_z_only_offset_too` (Z-only offset `(0,0,1.25)` + sidecar `(0,15,0)` raises, citing the Z component).
- **Escape hatch** — an explicit `"offset": [0,0,0]` sets `sidecar_offset` non-`None`, condition 2 fails, and it wins the merge with no code change. Verified end-to-end through JSON parsing, not just at the dataclass, by `test_the_documented_escape_hatch_unblocks_the_run` (rc 0, footprint patched). The error text names it as a remedy.
- **Diff confinement in `models3d.py`** — hunks at `:646`, `:718`, `:748-753` only. `add_model_refs_to_text` at `:835` is untouched, so #4586 rebases over this cleanly. The merge site is the intended ~3 lines: one helper call plus the corrected comment.
- **`CHANGELOG.md` untouched**, and `git diff --stat origin/main...HEAD -- boards/ CHANGELOG.md` is empty, satisfying the "no `boards/*/output/` change" criterion.
- **`LcscModelEntry` docstring flip** — accurate and adequate. The old sentence ("a sidecar that overrides only `rotate` does not suppress a packaged `offset`") is now only conditionally true, so asserting it unqualified would be a lie; the offset-only direction it was flipped to is unconditionally true, and the flip is not a dodge because the removed direction is restated immediately below with its guard, the escape hatch, and the issue reference. Both directions remain documented.

### Verification against current main

The PR was 2 commits behind main. `git merge-tree --write-tree origin/main` is clean, and I re-ran the full affected surface with `origin/main` merged into the PR branch inside an isolated worktree: `tests/test_lcsc_models.py` + `tests/test_pcb_add_3d_models.py` -> **202 passed**. `ruff check .` / `ruff format --check .` clean (the `MFR001` noqa warning is pre-existing on main and in an untouched file).

### CI

The initial run's `Board 03 End-to-End (LVS-only)` failure was infrastructure, not code — `curl: (7) Failed to connect to releases.astral.sh port 443` while installing `uv`, before any project code ran. Re-ran the failed job: **all 16 checks now pass**, `mergeStateStatus` is `CLEAN`.

### Notes (non-blocking, no action requested)

Exposure is still zero, as the issue states — `LCSC_MODEL_TRANSFORMS` has one `rotate`-only entry and both committed sidecars are bare-string form. `test_committed_sidecars_never_trip_the_guard` is the right way to keep AC 4 true under a future table addition rather than resting on today's shape. Exact float comparison is justified in the docstring on the correct grounds (authored literals; a sidecar writing `360` against a table's `0` has stated a rotation it never rendered) rather than merely excused.

LGTM.
